server/jetstream_api.go

```diff
@@ -1398,79 +1424,137 @@ func (s *Server) jsStreamRestoreRequest(sub *subscription, c *client, subject, r
-			s.Warnf("Storage failure for restore at %s for stream in account %q: %v",
-				FriendlyBytes(int64(received)), stream, acc.Name, err)
-			tfile.Close()
-			os.Remove(tfile.Name())
-			sub.client.processUnsub(sub.sid)
+			resultCh <- result{err, reply}
 			if reply != _EMPTY_ {
 				s.sendInternalAccountMsg(acc, reply, "-ERR 'storage failure during restore'")
 			}
 			return
 		}
-		received += len(msg)
+
+		activeCh <- len(msg)
 		s.sendInternalAccountMsg(acc, reply, nil)
-	})
+	}
 
+	sub, err := acc.subscribeInternal(restoreSubj, processChunk)
+	if err != nil {
+		resp.Error = &ApiError{Code: 500, Description: "JetStream unable to subscribe to restore snapshot"}
+		s.sendAPIResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+		return nil
+	}
+
+	// Mark the subject so the end user knows where to send the snapshot chunks.
 	resp.DeliverSubject = restoreSubj
 	s.sendAPIResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(resp))
+
+	doneCh := make(chan error, 1)
+
+	// Monitor the progress from another Go routine.
+	s.startGoRoutine(func() {
+		defer s.grWG.Done()
+		defer func() {
+			tfile.Close()
+			os.Remove(tfile.Name())
+			sub.client.processUnsub(sub.sid)
+		}()
+
+		const activityInterval = 2 * time.Second
+		notActive := time.NewTimer(activityInterval)
+		defer notActive.Stop()
+
+		total := 0
+		for {
+			select {
+			case result := <-resultCh:
+				end := time.Now()
+
+				// TODO(rip) - Should this have the error code in it??
+				s.publishAdvisory(acc, JSAdvisoryStreamRestoreCompletePre+"."+stream, &JSRestoreCompleteAdvisory{
+					TypedEvent: TypedEvent{
+						Type: JSRestoreCompleteAdvisoryType,
+						ID:   nuid.Next(),
+						Time: time.Now().UTC(),
+					},
+					Stream: stream,
+					Start:  start.UTC(),
+					End:    end.UTC(),
+					Bytes:  int64(total),
+					Client: ci,
+				})
+
+				var mset *Stream
+				var resp = JSApiStreamCreateResponse{ApiResponse: ApiResponse{Type: JSApiStreamCreateResponseType}}
+
+				err := result.err
+				if err == nil {
+					mset, err = acc.LookupStream(stream)
+				}
+				if err != nil {
+					resp.Error = jsError(err)
+					s.Warnf("Restore failed for %s for stream '%s > %s' in %v",
+						FriendlyBytes(int64(total)), stream, acc.Name, end.Sub(start))
+				} else {
+					resp.StreamInfo = &StreamInfo{Created: mset.Created(), State: mset.State(), Config: mset.Config()}
+					s.Noticef("Completed restore of %s for stream '%s > %s' in %v",
+						FriendlyBytes(int64(total)), stream, acc.Name, end.Sub(start))
+				}
+
+				// On the last EOF, send back the stream info or error status.
+				s.sendInternalAccountMsg(acc, result.reply, s.jsonResponse(&resp))
+				// Signal to the upper layers.
+				doneCh <- err
+				return
+			case n := <-activeCh:
+				total += n
+				notActive.Reset(activityInterval)
+			case <-notActive.C:
+				err := fmt.Errorf("restore for stream '%s > %s' is stalled", acc, stream)
+				s.Warnf(err.Error())
+				doneCh <- err
```

Comment:
Since you don't return here, if the timer fires again, this will block (unless something receives on that channel and empty it).

Reply:
Will add return, thanks.